src/ast/sanity.ts

```diff
@@ -0,0 +1,623 @@
+            checkDirectChildren(node, "vCondition", "vTrueExpression", "vFalseExpression");
+        } else if (node instanceof ElementaryTypeNameExpression) {
+            if (!(typeof node.typeName === "string")) {
+                checkDirectChildren(node, "typeName");
+            }
+        } else if (node instanceof FunctionCall) {
+            checkDirectChildren(node, "vExpression", "vArguments");
+        } else if (node instanceof FunctionCallOptions) {
+            checkDirectChildren(node, "vExpression", "vOptions");
+        } else if (node instanceof Identifier || node instanceof IdentifierPath) {
+            if (node.referencedDeclaration !== null && node.vReferencedDeclaration !== undefined) {
+                checkFieldAndVFieldMatch(node, "referencedDeclaration", "vReferencedDeclaration");
+                checkVFieldCtx(node, "vReferencedDeclaration", ctx);
+            }
+        } else if (node instanceof IndexAccess) {
+            checkDirectChildren(node, "vBaseExpression", "vIndexExpression");
+        } else if (node instanceof IndexRangeAccess) {
+            checkDirectChildren(node, "vBaseExpression", "vStartExpression", "vEndExpression");
+        } else if (node instanceof MemberAccess) {
+            if (node.referencedDeclaration !== null && node.vReferencedDeclaration !== undefined) {
+                checkFieldAndVFieldMatch(node, "referencedDeclaration", "vReferencedDeclaration");
+                checkVFieldCtx(node, "vReferencedDeclaration", ctx);
+            }
+            checkDirectChildren(node, "vExpression");
+        } else if (node instanceof NewExpression) {
+            checkDirectChildren(node, "vTypeName");
+        } else if (node instanceof TupleExpression) {
+            checkDirectChildren(node, "vOriginalComponents", "vComponents");
+        } else if (node instanceof UnaryOperation) {
+            checkVFieldCtx(node, "vSubExpression", ctx);
+            checkDirectChildren(node, "vSubExpression");
+        } else {
+            throw new Error(`Unknown ASTNode type ${node.constructor.name}`);
+        }
+    }
+}
+
+/**
+ * Check that a single SourceUnit has a sane structure. This checks that:
+ *  - All reachable nodes belong to the same context, have their parent/sibling set correctly.
+ *  - All number id properties of nodes point to a node in the same context.
+ *  - When a number property (e.g. `scope`) has a corresponding `v` prefixed property (e.g. `vScope`)
+ *    check that the number proerty corresponds to the id of the `v` prefixed property.
+ *  - Most 'v' properties point to direct children of a node.
+ *
+ * NOTE: While this code can be slightly slow, its meant to be used mostly in testing so its
+ * not performance critical.
+ */
+export function isSane(unit: SourceUnit, ctx: ASTContext): boolean {
+    try {
+        checkSane(unit, ctx);
+    } catch (e) {
+        if (e instanceof InsaneASTError) {
+            console.error(e);
+
+            return false;
+        }
```

Comment:
Do we intentionally silent the `InsaneASTError`? Or is it for debugging purposes?

Reply:
Yeah - that error is just for debugging purposes - to communicate to `isSane` (or any other consumer) what exactly was the problem. If you want to do something else other than silence then you call `checkSane` directly